rust/src/writer.rs

```diff
@@ -0,0 +1,254 @@
+//! The writer module contains the DeltaWriter and DeltaWriterError definitions and provides a
+//! higher level API for performing Delta transaction writes to a given Delta Table.
+//!
+//! Unlike the transaction API on DeltaTable, this higher level writer will also write out the
+//! parquet files
+
+use crate::action::Txn;
+use crate::{DeltaTableError, DeltaTransactionError};
+use arrow::record_batch::RecordBatch;
+use parquet::arrow::ArrowWriter;
+use parquet::basic::Compression;
+use parquet::file::properties::WriterProperties;
+use parquet::file::writer::InMemoryWriteableCursor;
+use serde_json::Value;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::sync::Arc;
+
+/**
+ * BufferedJSONWriter allows for buffering serde_json::Value rows before flushing to parquet files
+ * and a Delta transaction
+ */
+pub struct BufferedJSONWriter {
+    table: crate::DeltaTable,
+    buffer: HashMap<WriterPartition, Vec<Value>>,
+    schema: arrow::datatypes::SchemaRef,
+    partitions: Vec<String>,
+    txns: Vec<Txn>,
+}
+
+impl BufferedJSONWriter {
+    /**
+     * Attempt to construct the BufferedJSONWriter, will fail if the table's metadata is not
+     * present
+     */
+    pub fn try_new(table: crate::DeltaTable) -> Result<Self, DeltaTableError> {
+        let metadata = table.get_metadata()?.clone();
+        let schema = metadata.schema;
+        let arrow_schema =
+            <arrow::datatypes::Schema as TryFrom<&crate::Schema>>::try_from(&schema).unwrap();
+        let schema = Arc::new(arrow_schema);
+
+        Ok(Self {
+            table,
+            schema,
+            buffer: HashMap::new(),
+            partitions: metadata.partition_columns,
+            txns: vec![],
+        })
+    }
+
+    /**
+     * Return the total Values pending in the buffer
+     */
+    pub fn count(&self, partitions: &WriterPartition) -> Option<usize> {
+        if let Some(buffer) = self.buffer.get(&partitions) {
+            return Some(buffer.len());
+        }
+        None
```

Comment:
nit, we can probably simplify this to:
```suggestion
        self.buffer.get(&partitions).map(|b| b.len())
```